sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+//
+
+#include <sound/sof/ipc4/header.h>
+#include <sound/sof/xtensa.h>
+#include "../ipc4-priv.h"
+#include "../sof-priv.h"
+#include "hda.h"
+#include "telemetry.h"
+
+void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
+{
+	struct telemetry_slot_data *telemetry_data;
+	struct sof_ipc_dsp_oops_xtensa *xoops;
+	u32 slot_offset;
+	char *level;
+
+	slot_offset = sdev->debug_box.offset + (SOF_IPC4_DEBUG_TELEMETRY_SLOT_ID + 1) *
+				SOF_IPC4_DEBUG_SLOT_SIZE;
+
+	telemetry_data = kmalloc(sizeof(*telemetry_data), GFP_KERNEL);
+	if (!telemetry_data)
+		return;
+
+	xoops = kzalloc(sizeof(*xoops) + XTENSA_CORE_AR_REGS_COUNT * sizeof(int), GFP_KERNEL);
```

Comment:
```c
kzalloc(flex_array_size(xoops, ar, XTENSA_CORE_AR_REGS_COUNT), GFP_KERNEL);
```
and probably move it after the header checks?

Reply:
Good idea ! thanks